x/lockup/ante.go

```diff
@@ -64,81 +84,125 @@ func (lad LockAnteDecorator) AnteHandle(
+			// Since authz MsgExec holds Msgs inside of it, all those inner Msgs must be checked
 			case *authz.MsgExec:
-				// Check that all of the inner Msgs are not locked types
 				for _, m := range msg.Msgs {
 					var inner sdk.Msg
 					err := lad.cdc.UnpackAny(m, &inner)
 					if err != nil {
 						return ctx,
 							sdkerrors.Wrapf(sdkerrors.ErrInvalidType, "unable to unpack authz msgexec message: %v", err)
 					}
+					// Check if the inner Msg is acceptable or not, returning an error kicks this whole Tx out of the mempool
 					if err := lad.isAcceptable(ctx, inner); err != nil {
 						return ctx, err
 					}
 				}
+			// TODO: If ICA is used, those Msgs must be checked as well or they could cause a loophole for us
 			default:
-				// Check that this msg is not one of the locked types
+				// Check if this Msg is acceptable or not, returning an error kicks this whole Tx out of the mempool
 				if err := lad.isAcceptable(ctx, msg); err != nil {
 					return ctx, err
 				}
 			}
-
 		}
 	}
 
 	return next(ctx, tx, simulate)
 }
 
-// isAcceptable checks if the given msg is acceptable, returns an error if not
+// isAcceptable checks if the given msg is permissible under a locked chain, returns an error if not
+// A Msg is not permissibile if it involves the transfer of a locked token via a locked Msg type from a nonexempt address
+// All of these conditions are determined by what is stored in the lockup module params
 func (lad LockAnteDecorator) isAcceptable(ctx sdk.Context, msg sdk.Msg) error {
+	lockedTokenDenomsSet := lad.lockupKeeper.GetLockedTokenDenomsSet(ctx)
 	lockedMsgTypesSet := lad.lockupKeeper.GetLockedMessageTypesSet(ctx)
 	exemptSet := lad.lockupKeeper.GetLockExemptAddressesSet(ctx)
 	if _, typePresent := lockedMsgTypesSet[sdk.MsgTypeURL(msg)]; typePresent {
-		if allow, err := allowMessage(msg, exemptSet); !allow {
+		// Check that any locked msg is permissible on a type-case basis
+		if allow, err := allowMessage(msg, exemptSet, lockedTokenDenomsSet); !allow {
 			return sdkerrors.Wrap(err, fmt.Sprintf("Transaction blocked because of message %v", msg))
 		}
 	}
 
 	return nil
 }
 
-// NewAnteHandler returns an AnteHandler that ensures any transaction under a locked chain
-// originates from a LockExempt address
-func NewLockupAnteHandler(lockupKeeper keeper.Keeper, cdc codec.Codec) sdk.AnteHandler {
-	return sdk.ChainAnteDecorators(NewLockupAnteDecorator(lockupKeeper, cdc))
-}
-
-// NewLockupAnteDecorator initializes a LockupAnteDecorator for locking messages
-// based on the settings stored in lockupKeeper
-func NewLockupAnteDecorator(lockupKeeper keeper.Keeper, cdc codec.Codec) LockAnteDecorator {
-	return LockAnteDecorator{lockupKeeper, cdc}
-}
-
-// allowMessage checks that an input `msg` was sent by only addresses in `exemptSet`
-// returns true if `msg` is either permissible or not a type of message this module blocks
-func allowMessage(msg sdk.Msg, exemptSet map[string]struct{}) (bool, error) {
```

Comment:
Here I have added both the token blacklist and support for more Msg types: ibc transfers, microtx xfers, and bank send + multi send